api/desecapi/replication.py

```diff
@@ -0,0 +1,125 @@
+        self.path = path
+
+    def _git(self, *args):
+        cmd = ['/usr/bin/git'] + list(args)
+        print('>>> ' + str(cmd))
+
+        with subprocess.Popen(
+                cmd,
+                bufsize=0,
+                cwd=self.path,
+                stderr=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+        ) as p:
+            rcode = p.wait()
+            stderr = p.stderr.read()
+            stdout = p.stdout.read()
+            try:
+                stderr, stdout = stderr.decode(), stdout.decode()
+            except UnicodeDecodeError:
+                GitRepositoryException('git stdout or stderr was not valid unicode!', cmd=cmd, stderr=stderr, stdout=stdout)
+
+        if rcode != 0:
+            raise GitRepositoryException('git returned nonzero error code', cmd=cmd, rcode=rcode, stdout=stdout, stderr=stderr)
+
+        print('\n'.join(['<<< ' + s for s in stdout.split('\n')]))
+        return stdout
+
+    def commit_all(self, msg=None):
+        self._git('add', '.')
+        status = self._git('status')
+        if 'nothing to commit' not in status:
+            self._git('commit', '-m', msg or 'update')
+        else:
+            print('...')
+
+    def init(self, name='deSEC', email='git@desec.internal'):
+        self._git('init')
+        self._git('config', 'user.name', name)
+        self._git('config', 'user.email', email)
+        # TODO on commit, git tries to access $HOME/.config/git
+        #  but due to celery privilege drop, $HOME is still /root
+        #  and this gives a 'permission denied' warning in the log
+
+    def get_head(self):
+        return self.get_commit('HEAD')
+
+    def get_commit(self, rev):
+        try:
+            commit_hash = self._git('show', rev, '--format=%s', '-s')[:-1]
+            commit_message = self._git('show', rev, '--format=%H', '-s')[:-1]
+            return commit_hash, commit_message
+        except GitRepositoryException as e:
+            print('@@@@', e.stderr)
+            return None, None
+
+    def remove_history(self, before: timedelta):
+        rev = self._git('log', f'--before={(datetime.now() - before).isoformat()}', '-1', '--format=%H')
+        with open(os.path.join(self.path, '.git', 'shallow'), 'w') as f:
+            f.writelines([rev])
+        self._git('reflog', 'expire', '--expire=now', '--all')
+        self._git('gc', '--prune=now')
+
+
+class ZoneRepository(Repository):
+
+    def __init__(self, path):
+        super().__init__(path)
+        if not os.path.exists(os.path.join(self.path, '.git')):
+            self.init()
+            self.commit_all(msg='Inception or Recovery')
+            self.full_refresh()
+            self.commit_all(msg='Full Refresh')
+
+    def full_refresh(self):
+        print('WARNING: Full refresh of replication zone repository!')
+        for name in models.Domain.objects.values_list('name', flat=True):
+            self.axfr(name)
+
+    def axfr(self, name, ns='172.16.1.11'):
```

Comment:
I assumed this would trigger an AXFR to some frontend server. Maybe call this `export` or `dump`? (It's also not important that it causes an AXFR query; we might do something else in the future, e.g. query the pdns API.)